File: sdk/src/dial.ts
/**
 * The shipped **dial** kind — temper's own, not a provider's.
 *
 * A local-locus TOML document at `.temper/dial.toml` whose entries name a clause by
 * its compiled address (`contract.md`, "clause") and declare the severity this machine
 * reads it at: see the annoying finding, read its label, dial it — locally,
 * uncommitted, and without ever softening the shared gate.
 *
 * This module is **root-exported** (`@dtmd/temper`), never the `./claude-code`
 * subpath: the subpath is the Claude Code provider face, and the dial is a claim about
 * temper's own gate rather than about any harness Claude Code reads. Its facts are
 * therefore not external facts and carry no `cite` — the citation is the corpus
 * (decisions 0030, 0032).
 *
 * **The schema is the envelope** (decision 0030, re-spelled by 0032). Severity is the
 * only verb an entry has: there is no `off`, no `skip`, no `delete`, and the severity
 * vocabulary is the same closed two the authored clause declares under. A dialed clause
 * therefore still reports, and the bound 0030 stated as an admissibility condition over
 * a layer's *effect* is held structurally — by the shape of a kind, which an author
 * cannot spell their way past. The other half of the bound is the engine's:
 * dialed softening is inert in block mode, so a block-mode pass on any machine implies
 * the shared gate's pass (`pipeline.md`, "Layers").
 */

import { clause, closedKeys, enumOf, required, type } from "./contract.js";
import type { Clause, Severity } from "./contract.js";
import { kind } from "./kind.js";
import type { KindDefinition } from "./kind.js";

/**
 * One dial entry: a clause's compiled address, and the severity this machine reads it
 * at. The two fields are the whole verb — the interface is where the envelope is
 * spelled, and there is nothing here to widen it with.
 */
export interface DialEntry {
  /**
   * The clause's compiled address — the label every finding prints as its `rule` id
   * and `explain` narrates (decision 0032). Spelled straight back out of the finding:
   * that round trip is what the label's legibility exists for.
   */
  readonly label: string;
  /** The severity this machine reads that clause at. */
  readonly severity: Severity;
}

/**
 * The dial document's fields — the whole top-level table of `.temper/dial.toml`.
 *
 * `name` is the machine's own label, and it is the member's identity: the file's stem is
 * `dial` on every machine that has one, so identity comes from a declared field.
 */
export interface Dial {
  /** This machine's name — the dial's identity, and what its findings speak. */
  readonly name: string;
  /** The dialed clauses. Absent or empty is a dial that changes nothing. */
  readonly clause?: readonly DialEntry[];
}

/**
 * The `dial` kind: `.temper/dial.toml`, a **local** file locus — the kind is declared,
 * committed and reviewed here; a machine's own dial document is not.
 *
 * Read-side only by construction (decision 0034): `toml-document` is a read face with no
 * write twin, emit's codomain is the committed tree, and a local member's rows never
 * enter the lock. Channel-less: a dial is read by temper's own gate, never surfaced to a
 * model or an installer, so it declares no registration.
 */
export const dial: KindDefinition<Dial> = kind<Dial>({
  name: "dial",
  locus: { kind: "at", root: ".temper", glob: "dial.toml", commitment: "local" },
  format: "toml-document",
  unitShape: "named-field",
  registration: [],
  identityField: "name",
});

/**
 * The default contract for `dial` — the envelope, stated as clauses.
 *
 * Every clause here is `required`: this contract is the one thing on the machine a dial
 * must not be able to soften, since softening it is how a dial would talk its way out of
 * its own shape. The engine refuses a dial entry naming a `dial.*` label for that reason
 * — the clauses below are what an entry is *checked against*, never what it ranges over.
 *
 * `closed-keys` is the load-bearing one: it is what makes a misspelt verb a finding
 * instead of a silently ignored key. A dial that wrote `off = true` beside its entries
 * would otherwise read as an honest dial that happened to do nothing.
 */
export const dialDefaultContract: readonly Clause[] = [
  clause(required("name"), {
    severity: "required",
    guidance:
      "A dial names the machine it speaks for — its identity, and what its findings are reported under. There is no filename fallback: every machine's dial is `dial.toml`.",
  }),
  clause(type("name", ["string"]), { severity: "required" }),
  clause(type("clause", ["list"]), {
    severity: "required",
    guidance:
      "The dialed clauses are an array of tables — `[[clause]]` per entry, each naming one `label` and one `severity`.",
  }),
  clause(required("clause[*].label"), {
    severity: "required",
    guidance:
      "An entry names the clause it dials by the address the finding printed as its `rule` id — spell it back verbatim. An entry naming a label no clause in this harness carries is a finding, never a silent no-op.",
  }),
  clause(type("clause[*].label", ["string"]), { severity: "required" }),
  clause(required("clause[*].severity"), {
    severity: "required",
    guidance:
      "Severity is an entry's only verb: it is what a dial says, and all a dial says.",
  }),
  clause(enumOf("clause[*].severity", ["required", "advisory"]), {
    severity: "required",
    guidance:
      "The two severities an authored clause declares under are the two a dial re-declares under — there is no third value that means `off`. Softening a clause to `advisory` still reports it, and is inert in block mode; hardening one to `required` binds in every mode.",
  }),
  clause(closedKeys(), {
    severity: "required",
    guidance:
      "The dial's key set is exhaustive: `name` and `clause`, nothing else. A key outside it is a verb this schema does not have — most likely an attempt to spell deletion, which it cannot.",
  }),
];

File: src/dial.rs
//! The shipped `dial` kind's read side: the severities this machine reads clauses at.
//!
//! `.temper/dial.toml` names clauses by the compiled address every finding already
//! prints, and declares a severity per entry. Severity is the only verb the schema has
//! — [`crate::builtin_kind`] declares the kind, `dialDefaultContract` closes its key set
//! — so deletion is unspellable and a dialed clause still reports.
//!
//! Two bounds hold here rather than in the schema, because neither is a shape.

use std::collections::{BTreeMap, BTreeSet};

use serde_json::Value as JsonValue;

use crate::check::Diagnostic;
use crate::compose::{self, EnforcementMode};
use crate::contract::{Clause, Severity};
use crate::extract::Features;

/// The kind's bare name — temper's own, and the owner segment of every label it may not
/// name ([`Dial::refusals`]).
pub const KIND: &str = "dial";

/// The document the kind governs, under the workspace root.
pub const DOCUMENT: &str = "dial.toml";

/// The top-level key the entries sit under: `[[clause]]`.
const ENTRIES_KEY: &str = "clause";

/// The rule id of a finding about a dial *entry* — never a clause label, so it is
/// undialable, which is the point: an entry that says nothing must not be able to
/// silence the report that says so.
const ENTRY_RULE: &str = "dial.entry";

/// This machine's declared severities, keyed by the clause address each entry names.
///
/// An entry the schema would reject never lands here — a missing or non-string label, a
/// severity outside the closed two — because failing to apply is the safe direction and
/// the kind's own contract is what reports it. Silence is impossible: the entry survives
/// into [`Dial::refusals`] as a label nothing matched.
#[derive(Debug, Clone, Default)]
pub struct Dial {
    entries: BTreeMap<String, Severity>,
}

impl Dial {
    /// This machine's dial, read off the `dial` kind's extracted members.
    ///
    /// More than one dial document cannot arise — the kind governs one glob with no
    /// wildcard — but the read folds every member rather than assuming it, and a label
    /// two members share resolves to the last one folded.
    #[must_use]
    pub fn from_features(members: &[Features]) -> Self {
        let mut entries = BTreeMap::new();
        for member in members {
            let Some(JsonValue::Array(rows)) = member.fields.get(ENTRIES_KEY) else {
                continue;
            };
            for row in rows {
                let (Some(JsonValue::String(label)), Some(JsonValue::String(severity))) =
                    (row.get("label"), row.get("severity"))
                else {
                    continue;
                };
                if let Some(severity) = compose::severity_from_label(severity) {
                    entries.insert(label.clone(), severity);
                }
            }
        }
        Self { entries }
    }

    /// Whether this machine dials nothing — the common case, and the one every caller
    /// below is a no-op for.
    #[must_use]
    #[allow(dead_code)]
    fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }

    /// Re-read every clause in `clauses` this dial names, at the severity it declares —
    /// returning the addresses it actually reached.
    ///
    /// Hardening (advisory → required) binds under every mode. Softening is the reviewed
    /// half: `mode` [`Block`](EnforcementMode::Block) leaves the authored severity
    /// standing, so the machine can only ever be stricter than the shared gate, never
    /// laxer. A clause the dial re-declares at the severity it already carries is neither,
    /// and applies silently.
    ///
    /// Idempotent, and called over overlapping clause sets on purpose: a set-grain clause
    /// reaches the judge through a selection assembled out of contracts this has already
    /// run over.
    pub fn apply(&self, mode: EnforcementMode, clauses: &mut [Clause]) -> BTreeSet<String> {
        let mut reached = BTreeSet::new();
        for clause in clauses.iter_mut() {
            let Some(&severity) = self.entries.get(&clause.label) else {
                continue;
            };
            reached.insert(clause.label.clone());
            let softens = clause.severity == Severity::Required && severity == Severity::Advisory;
            if softens && mode == EnforcementMode::Block {
                continue;
            }
            clause.severity = severity;
        }
        reached
    }

    /// Every entry that dialed nothing, as a finding — the two ways a dial can be wrong
    /// about the gate rather than malformed against its own schema.
    ///
    /// Both are errors, and both are strictly *stricter* than the shared gate: a machine
    /// that fails its own check over a stale dial has still never softened anything. The
    /// alternative — an entry that quietly names nothing — is the outcome the address's
    /// legibility exists to prevent.
    #[must_use]
    pub fn refusals(&self, reached: &BTreeSet<String>) -> Vec<Diagnostic> {
        let own_prefix = format!("{KIND}.");
        self.entries
            .keys()
            .filter(|label| !reached.contains(label.as_str()))
            .map(|label| {
                let message = if label.starts_with(&own_prefix) {
                    format!(
                        "dial entry `{label}` names the dial's own contract, which no dial \
                         may re-read: that contract is the envelope this document is \
                         checked against"
                    )
                } else {
                    format!(
                        "dial entry `{label}` names no clause this harness carries — spell \
                         the address a finding printed as its `rule` id, or drop the entry"
                    )
                };
                Diagnostic::error(ENTRY_RULE, DOCUMENT, message)
            })
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// One dial member carrying `entries` under the entries key, and nothing else — the
    /// only feature this module's read looks at.
    fn dial_member(entries: Option<JsonValue>) -> Features {
        Features {
            id: "workstation".to_string(),
            fields: entries
                .into_iter()
                .map(|value| (ENTRIES_KEY.to_string(), value))
                .collect(),
            body_lines: 0,
            rendered_lines: Some(0),
            rendered_chars: Some(0),
            headings: Vec::new(),
            sections: Vec::new(),
            source_dir: None,
            directives: Vec::new(),
            fenced_blocks: Vec::new(),
            nested_members: Vec::new(),
            satisfies: Vec::new(),
            edge_placements: None,
        }
    }

    #[test]
    fn an_entry_outside_the_severity_vocabulary_never_lands() {
        // Failing to apply is the safe direction, and the kind's own `enum` clause is
        // what reports it — the read never invents a third severity's meaning.
        let dial = Dial::from_features(&[dial_member(Some(serde_json::json!([
            { "label": "skill.extent", "severity": "off" },
            { "label": "rule.required.description", "severity": "advisory" },
        ])))]);

        assert_eq!(
            dial.entries.keys().collect::<Vec<_>>(),
            vec!["rule.required.description"]
        );
    }

    #[test]
    fn an_entry_missing_either_half_never_lands() {
        let dial = Dial::from_features(&[dial_member(Some(serde_json::json!([
            { "label": "skill.extent" },
            { "severity": "advisory" },
            { "label": ["skill.extent"], "severity": "advisory" },
        ])))]);

        assert!(dial.is_empty());
    }

    #[test]
    fn a_dial_declaring_no_entries_at_all_is_empty() {
        assert!(Dial::from_features(&[dial_member(None)]).is_empty());
    }
}
